AK/CircularBuffer.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2022, Lucas Chollet <lucas.chollet@free.fr>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/ByteBuffer.h>
+#include <AK/Error.h>
+#include <AK/Noncopyable.h>
+
+namespace AK {
+
+class CircularBuffer {
+    AK_MAKE_NONCOPYABLE(CircularBuffer);
+
+public:
+    static ErrorOr<CircularBuffer> create_empty(size_t size);
+    static ErrorOr<CircularBuffer> create_initialized(ByteBuffer);
+
+    CircularBuffer(CircularBuffer&& other)
+    {
+        operator=(move(other));
+    }
+
+    CircularBuffer& operator=(CircularBuffer&& other)
+    {
+        if (&other == this)
+            return *this;
+
+        swap(m_buffer, other.m_buffer);
+        swap(m_reading_head, other.m_reading_head);
+        swap(m_used_space, other.m_used_space);
+
+        return *this;
+    }
```

Comment:
```suggestion
CircularBuffer(CircularBuffer&&) = default;
CircularBuffer& operator=(CircularBuffer&&) = default;
```

These are exactly the same as a defaulted implementation :)